src/source/Response.c

```diff
@@ -625,25 +655,76 @@ SIZE_T postReadCallback(PCHAR pBuffer, SIZE_T size, SIZE_T numItems, PVOID custo
     pCurlApiCallbacks = pCurlRequest->pCurlApiCallbacks;
     uploadHandle = pCurlResponse->pCurlRequest->uploadHandle;
 
-    if (pCurlResponse->paused) {
-        bytesWritten = CURL_READFUNC_PAUSE;
+    if (pCurlResponse->endOfStream || ATOMIC_LOAD_BOOL(&pCurlRequest->requestInfo.terminating)) {
+        DLOGD("Closing connection for upload stream handle: %"
+        PRIu64, uploadHandle);
         CHK(FALSE, retStatus);
     }
 
-    if (pCurlResponse->endOfStream || ATOMIC_LOAD_BOOL(&pCurlRequest->requestInfo.terminating)) {
-        DLOGI("Closing connection for upload stream handle: %" PRIu64, uploadHandle);
-        CHK(FALSE, retStatus);
+    pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.lockMutexFn(
+            pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.customData,
+            pCurlResponse->dataAvailableLock);
+
+    while (bytesWritten == 0 && !ATOMIC_LOAD_BOOL(&pCurlResponse->dataAvailable)) {
+        awaitStatus = CVAR_WAIT(pCurlResponse->dataAvailableCvar, pCurlResponse->dataAvailableLock,
+                                DEFAULT_LOW_SPEED_TIME_LIMIT);
+        if (awaitStatus == STATUS_OPERATION_TIMED_OUT) {
+            DLOGW("Time out...");
+            retStatus = awaitStatus;
+            break;
+        } else if (!ATOMIC_LOAD_BOOL(&pCurlResponse->dataAvailable)) {
+            DLOGW("Spurious wakeup...");
+            // spurious wakeup
+            continue;
+        } else {
+            retStatus = getKinesisVideoStreamData(pCurlResponse->pCurlRequest->streamHandle, uploadHandle, (PBYTE) pBuffer,
+                                                  (UINT32) bufferSize, &retrievedSize);
+
+            if (pCurlApiCallbacks->curlReadCallbackHookFn != NULL) {
+                retStatus =
+                        pCurlApiCallbacks->curlReadCallbackHookFn(pCurlResponse, uploadHandle, (PBYTE) pBuffer,
+                                                                  (UINT32) bufferSize, &retrievedSize, retStatus);
+            }
+
+            DLOGW("Real wakeup.... bytes Written should NOT be zero");
+            bytesWritten = (SIZE_T) retrievedSize;
+
+            if (bytesWritten > 0) {
+                ATOMIC_STORE_BOOL(&pCurlResponse->dataAvailable, FALSE);
+            }
+
+            if(retStatus == STATUS_AWAITING_PERSISTED_ACK) {
+                continue;
+            } else {
+                break;
+            }
+        }
     }
 
-    retStatus =
-        getKinesisVideoStreamData(pCurlResponse->pCurlRequest->streamHandle, uploadHandle, (PBYTE) pBuffer, (UINT32) bufferSize, &retrievedSize);
 
-    if (pCurlApiCallbacks->curlReadCallbackHookFn != NULL) {
-        retStatus =
-            pCurlApiCallbacks->curlReadCallbackHookFn(pCurlResponse, uploadHandle, (PBYTE) pBuffer, (UINT32) bufferSize, &retrievedSize, retStatus);
+    if (ATOMIC_LOAD_BOOL(&pCurlResponse->dataAvailable)) {

```

Comment:
Why do you need to call getKinesisVideoStreamData again? As I recommended, you could handle everything in a single while loop